Duplicate local variable name `bitmap$0` on 2.10/2.11 for some weird lazy vals

Discovered in #3178.

2 partests fail to IR-check: [run/t6272.scala](https://github.com/scala/scala/blob/v2.12.4/test/files/run/t6272.scala) and [run/t3877.scala](https://github.com/scala/scala/blob/v2.12.4/test/files/run/t3877.scala), on 2.11 (and also if ported to 2.10). On 2.12+, the produced IR is correct. Given that they are caused by `lazy` vals, this is not surprising, since scalac significant changed its treatment of lazy vals in 2.12. The errors (for t6272) are:
```
[error] Test.scala(30:7:VarDef): Duplicate local variable name bitmap$0.
[error] Test.scala(2:10:VarDef): Duplicate local variable name bitmap$0.
```
The issue is reproducible in 0.6.x and on master.

The IR looks like this:
```scala
  def init___() {
    this.O::init___();
    mod:LA$<-this;
    val bitmap$0: sr_VolatileByteRef = new sr_VolatileByteRef().init___B(0);
    try {
      val x1$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x1$1__p1__sr_IntRef__sr_VolatileByteRef__I(x1$lzy, bitmap$0));
      mod:s_sys_package$.error__T__sr_Nothing$("!")
    } catch (e) {
      val e$2: jl_Throwable = mod:sjsr_package$.wrapJavaScriptException__O__jl_Throwable(e);
      val x2$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x2$1__p1__sr_IntRef__sr_VolatileByteRef__I(x2$lzy, bitmap$0))
    } finally {
      val x3$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x3$1__p1__sr_IntRef__sr_VolatileByteRef__I(x3$lzy, bitmap$0))
    };
    if (mod:sjsr_RuntimeString$.isEmpty__T__Z("")) {
      val x4$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x4$1__p1__sr_IntRef__sr_VolatileByteRef__I(x4$lzy, bitmap$0))
    };
    this.b$1 = true;
    while (this.b__Z()) {
      val bitmap$0: sr_VolatileByteRef = new sr_VolatileByteRef().init___B(0);
      val x5$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x5$1__p1__sr_IntRef__sr_VolatileByteRef__I(x5$lzy, bitmap$0));
      this.b$und$eq__Z__V(false)
    }
  }
```
and
```scala
  def method__V() {
    val bitmap$0: sr_VolatileByteRef = new sr_VolatileByteRef().init___B(0);
    try {
      val x6$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x6$1__p1__sr_IntRef__sr_VolatileByteRef__I(x6$lzy, bitmap$0));
      mod:s_sys_package$.error__T__sr_Nothing$("!")
    } catch (e) {
      val e$2: jl_Throwable = mod:sjsr_package$.wrapJavaScriptException__O__jl_Throwable(e);
      val x7$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x7$1__p1__sr_IntRef__sr_VolatileByteRef__I(x7$lzy, bitmap$0))
    } finally {
      val x8$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x8$1__p1__sr_IntRef__sr_VolatileByteRef__I(x8$lzy, bitmap$0))
    };
    if (mod:sjsr_RuntimeString$.isEmpty__T__Z("")) {
      val x9$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x9$1__p1__sr_IntRef__sr_VolatileByteRef__I(x9$lzy, bitmap$0))
    };
    var b: boolean = true;
    while (b) {
      val bitmap$0: sr_VolatileByteRef = new sr_VolatileByteRef().init___B(0);
      val x10$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x10$1__p1__sr_IntRef__sr_VolatileByteRef__I(x10$lzy, bitmap$0));
      b = false
    }
  }
```

The code produced by scalac is actually wrong even on the JVM, as it does the wrong thing in some corner cases:
```scala
> scala
Welcome to Scala 2.11.11 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_60).
Type in expressions for evaluation. Or try :help.

scala> def method(): Unit = {
     |   var x1InitCount = 0
     |   lazy val x1 = { x1InitCount += 1; 1 }
     |   assert(x1InitCount == 0)
     |   println(x1)
     |   assert(x1InitCount == 1)
     |
     |   do {
     |     lazy val x2 = 2
     |     assert(x1InitCount == 1)
     |     println(x2)
     |     assert(x1InitCount == 1)
     |   } while (false)
     |
     |   println(x1)
     |   assert(x1InitCount == 1)
     | }
method: ()Unit

scala> method()
1
2
1
java.lang.AssertionError: assertion failed
  at scala.Predef$.assert(Predef.scala:156)
  at .method(<console>:26)
  ... 32 elided
```
The last assertion fails, because the `bitmap$0` has been reset to `0` when entering the `do` loop, because of the redefinition of `bitmap$0`. This causes `x1` to be reinitialized on the last access.

So it does seem that, the way this is compiled on the JVM, it behaves as if `bitmap$0` was declared once at the beginning of the method, and then each `ValDef` declaring `bitmap$0` is like an `Assign` instead.

I guess we could reproduce this behavior in our back-end as follows: in 2.10/2.11, we track which local vars have been declared. If we declare a variable twice in a method, we schedule the IR of that method for a fixup round, where we 1) declare all such variables as mutable at the beginning of the method, initialized with the zero of their type and 2) turn all their `ValDef`s into `Assign`s. It will incur some performance overhead even for sane methods, though, because of the tracking.

@gzm0 WDYT?

--- a/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.0/BlacklistedTests.txt
+++ b/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.0/BlacklistedTests.txt
@@ -870,6 +870,10 @@ run/t3493.scala
 # Using Class.forName
 run/private-inline.scala
 
+# Suffers from #3180
+run/t3877.scala
+run/t6272.scala
+
 ###  Incorrect partests  ###
 # Badly uses constract of Console.print (no flush)
 run/t429.scala

--- a/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.0/WhitelistedTests.txt
+++ b/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.0/WhitelistedTests.txt
@@ -1025,7 +1025,6 @@ run/t6337a.scala
 run/exoticnames.scala
 run/t0936.scala
 run/runtime-richChar.scala
-run/t6272.scala
 run/t7215.scala
 run/t1939.scala
 run/ReverseSeqView.scala
@@ -1278,7 +1277,6 @@ run/t0042.scala
 run/t3050.scala
 run/t4536.scala
 run/NestedClasses.scala
-run/t3877.scala
 run/seqlike-kmp.scala
 run/t5907.scala
 run/t266.scala

--- a/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.1/BlacklistedTests.txt
+++ b/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.1/BlacklistedTests.txt
@@ -878,6 +878,10 @@ run/t3493.scala
 # Using Class.forName
 run/private-inline.scala
 
+# Suffers from #3180
+run/t3877.scala
+run/t6272.scala
+
 ###  Incorrect partests  ###
 # Badly uses constract of Console.print (no flush)
 run/t429.scala

--- a/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.1/WhitelistedTests.txt
+++ b/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.1/WhitelistedTests.txt
@@ -1025,7 +1025,6 @@ run/t6337a.scala
 run/exoticnames.scala
 run/t0936.scala
 run/runtime-richChar.scala
-run/t6272.scala
 run/t7215.scala
 run/t1939.scala
 run/ReverseSeqView.scala
@@ -1278,7 +1277,6 @@ run/t0042.scala
 run/t3050.scala
 run/t4536.scala
 run/NestedClasses.scala
-run/t3877.scala
 run/seqlike-kmp.scala
 run/t5907.scala
 run/t266.scala

--- a/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.11/BlacklistedTests.txt
+++ b/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.11/BlacklistedTests.txt
@@ -988,6 +988,10 @@ run/t3493.scala
 # Using Class.forName
 run/private-inline.scala
 
+# Suffers from #3180
+run/t3877.scala
+run/t6272.scala
+
 ###  Incorrect partests  ###
 # Badly uses constract of Console.print (no flush)
 run/t429.scala

--- a/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.11/WhitelistedTests.txt
+++ b/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.11/WhitelistedTests.txt
@@ -1118,7 +1118,6 @@ run/t6337a.scala
 run/exoticnames.scala
 run/t0936.scala
 run/runtime-richChar.scala
-run/t6272.scala
 run/t7215.scala
 run/t1939.scala
 run/ReverseSeqView.scala
@@ -1368,7 +1367,6 @@ run/t0042.scala
 run/t3050.scala
 run/t4536.scala
 run/NestedClasses.scala
-run/t3877.scala
 run/seqlike-kmp.scala
 run/t5907.scala
 run/t266.scala

--- a/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.2/BlacklistedTests.txt
+++ b/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.2/BlacklistedTests.txt
@@ -892,6 +892,10 @@ run/t3493.scala
 # Using Class.forName
 run/private-inline.scala
 
+# Suffers from #3180
+run/t3877.scala
+run/t6272.scala
+
 ###  Incorrect partests  ###
 # Badly uses constract of Console.print (no flush)
 run/t429.scala

--- a/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.2/WhitelistedTests.txt
+++ b/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.2/WhitelistedTests.txt
@@ -1025,7 +1025,6 @@ run/t6337a.scala
 run/exoticnames.scala
 run/t0936.scala
 run/runtime-richChar.scala
-run/t6272.scala
 run/t7215.scala
 run/t1939.scala
 run/ReverseSeqView.scala
@@ -1278,7 +1277,6 @@ run/t0042.scala
 run/t3050.scala
 run/t4536.scala
 run/NestedClasses.scala
-run/t3877.scala
 run/seqlike-kmp.scala
 run/t5907.scala
 run/t266.scala

--- a/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.5/BlacklistedTests.txt
+++ b/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.5/BlacklistedTests.txt
@@ -919,6 +919,10 @@ run/t3493.scala
 # Using Class.forName
 run/private-inline.scala
 
+# Suffers from #3180
+run/t3877.scala
+run/t6272.scala
+
 ###  Incorrect partests  ###
 # Badly uses constract of Console.print (no flush)
 run/t429.scala

--- a/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.5/WhitelistedTests.txt
+++ b/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.5/WhitelistedTests.txt
@@ -1088,7 +1088,6 @@ run/t6337a.scala
 run/exoticnames.scala
 run/t0936.scala
 run/runtime-richChar.scala
-run/t6272.scala
 run/t7215.scala
 run/t1939.scala
 run/ReverseSeqView.scala
@@ -1338,7 +1337,6 @@ run/t0042.scala
 run/t3050.scala
 run/t4536.scala
 run/NestedClasses.scala
-run/t3877.scala
 run/seqlike-kmp.scala
 run/t5907.scala
 run/t266.scala

--- a/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.6/BlacklistedTests.txt
+++ b/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.6/BlacklistedTests.txt
@@ -921,6 +921,10 @@ run/t3493.scala
 # Using Class.forName
 run/private-inline.scala
 
+# Suffers from #3180
+run/t3877.scala
+run/t6272.scala
+
 ###  Incorrect partests  ###
 # Badly uses constract of Console.print (no flush)
 run/t429.scala

--- a/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.6/WhitelistedTests.txt
+++ b/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.6/WhitelistedTests.txt
@@ -1100,7 +1100,6 @@ run/t6337a.scala
 run/exoticnames.scala
 run/t0936.scala
 run/runtime-richChar.scala
-run/t6272.scala
 run/t7215.scala
 run/t1939.scala
 run/ReverseSeqView.scala
@@ -1350,7 +1349,6 @@ run/t0042.scala
 run/t3050.scala
 run/t4536.scala
 run/NestedClasses.scala
-run/t3877.scala
 run/seqlike-kmp.scala
 run/t5907.scala
 run/t266.scala

--- a/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.7/BlacklistedTests.txt
+++ b/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.7/BlacklistedTests.txt
@@ -949,6 +949,10 @@ run/t3493.scala
 # Using Class.forName
 run/private-inline.scala
 
+# Suffers from #3180
+run/t3877.scala
+run/t6272.scala
+
 ###  Incorrect partests  ###
 # Badly uses constract of Console.print (no flush)
 run/t429.scala

--- a/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.7/WhitelistedTests.txt
+++ b/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.7/WhitelistedTests.txt
@@ -1119,7 +1119,6 @@ run/t6337a.scala
 run/exoticnames.scala
 run/t0936.scala
 run/runtime-richChar.scala
-run/t6272.scala
 run/t7215.scala
 run/t1939.scala
 run/ReverseSeqView.scala
@@ -1369,7 +1368,6 @@ run/t0042.scala
 run/t3050.scala
 run/t4536.scala
 run/NestedClasses.scala
-run/t3877.scala
 run/seqlike-kmp.scala
 run/t5907.scala
 run/t266.scala

--- a/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.8/BlacklistedTests.txt
+++ b/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.8/BlacklistedTests.txt
@@ -966,6 +966,10 @@ run/t3493.scala
 # Using Class.forName
 run/private-inline.scala
 
+# Suffers from #3180
+run/t3877.scala
+run/t6272.scala
+
 ###  Incorrect partests  ###
 # Badly uses constract of Console.print (no flush)
 run/t429.scala

--- a/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.8/WhitelistedTests.txt
+++ b/partest-suite/src/test/resources/scala/tools/partest/scalajs/2.11.8/WhitelistedTests.txt
@@ -1118,7 +1118,6 @@ run/t6337a.scala
 run/exoticnames.scala
 run/t0936.scala
 run/runtime-richChar.scala
-run/t6272.scala
 run/t7215.scala
 run/t1939.scala
 run/ReverseSeqView.scala
@@ -1368,7 +1367,6 @@ run/t0042.scala
 run/t3050.scala
 run/t4536.scala
 run/NestedClasses.scala
-run/t3877.scala
 run/seqlike-kmp.scala
 run/t5907.scala
 run/t266.scala

--- a/partest/src/main/scala/scala/tools/nsc/MainGenericRunner.scala
+++ b/partest/src/main/scala/scala/tools/nsc/MainGenericRunner.scala
@@ -65,6 +65,7 @@ class MainGenericRunner {
         command.thingToRun, "main", command.arguments))
 
     val linkerConfig = StandardLinker.Config()
+      .withCheckIR(true)
       .withSemantics(semantics)
       .withSourceMap(false)
       .withOptimizer(optMode != NoOpt)

